### PR TITLE
feature(scheduler): implement plugin_evaluation_total metric

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

New ~"schedule_attempts"~ "plugin_evaluation_total" metric is added to the scheduler. This metric counts how many times the specific plugin affects the scheduling result while taking Skip status into consideration.

It's now only supporting the Filter status. After https://github.com/kubernetes/kubernetes/issues/114827, we can support "Score" as well.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/110643

#### Special notes for your reviewer:

Not sure if "schedule_attempts" is the most suitable name. Tell me if any other good idea on naming.

EDIT: renamed to `plugin_evaluation_total` as suggested in https://github.com/kubernetes/kubernetes/pull/115082#discussion_r1088025144.

#### Does this PR introduce a user-facing change?

```release-note
New "plugin_evaluation_total" is added to the scheduler. 
This metric counts how many times the specific plugin affects the scheduling result. The metric doesn't get incremented when the plugin has nothing to do with an incoming Pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/metrics/metrics.go [hunk 1]
```diff
@@ -189,8 +189,16 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		}, []string{"plugin", "profile"})
 
+	ScheduleAttempts = metrics.NewCounterVec(
```
Comment (alculquicondor) [inline pkg/scheduler/metrics/metrics.go:192]: counters should always end in `_total` https://prometheus.io/docs/practices/naming/#metric-names

`plugin_evaluation_total`? This is a low level metric, so I think it's fair to have the word plugin in it.

**Comment by alculquicondor** [comment]
@kerthcet or @chendave do you have cycles to look at this?

**Comment by kerthcet** [comment]
I'll be back tomorrow.
/assign

File: pkg/scheduler/metrics/metrics.go (hunk 1 above)
Comment (sanposhiho) [reply pkg/scheduler/metrics/metrics.go]: Renamed as suggested, thanks.

**Comment by sanposhiho** [comment]
@alculquicondor @kerthcet 

Fixed based on the suggestion, please take a look. 🙏

**Comment by kerthcet** [comment]
> This metric counts how many times the specific plugin affects the scheduling result while taking Skip status into consideration. In other words, the metric doesn't get incremented when the plugin has nothing to do with an incoming Pod.  

By evaluating the plug being called times, what can we get from this? I mean in what scenario we need this metrics. Not opposing it, something I haven't figured out.

**Comment by kerthcet** [comment]
And sorry for let this sitting a long time. Busy with works. ☹️

**Comment by alculquicondor** [comment]
> By evaluating the plug being called times, what can we get from this?

This is a common question in PRR for any new features: how can an admin know if a feature is being used? This PR moves the needle in that direction.

File: pkg/scheduler/framework/runtime/framework.go [hunk 2]
```diff
@@ -734,6 +734,7 @@ func (f *frameworkImpl) RunFilterPlugins(
 		if state.SkipFilterPlugins.Has(pl.Name()) {
 			continue
 		}
+		metrics.PluginEvaluationTotal.WithLabelValues(pl.Name(), Filter).Inc()
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:737]: Let's add PreFilter too? Just in case some plugins only implement PreFilter.

**Comment by kerthcet** [comment]
> This is a common question in PRR for any new features: how can an admin know if a feature is being used? This PR moves the needle in that direction.

Got it, before without the support of Skip, all enabled filter plugins' being called times should be equal, now they're not very likely, this can be grabbed by metrics.

File: pkg/scheduler/metrics/metrics.go [hunk 3]
```diff
@@ -189,6 +189,14 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		}, []string{"plugin", "profile"})
 
+	PluginEvaluationTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      SchedulerSubsystem,
+			Name:           "plugin_evaluation_total",
+			Help:           "Number of attempts to schedule pods by each plugin and the extension point (only Filter is supported now.).",
+			StabilityLevel: metrics.ALPHA,
+		}, []string{"plugin", "extension_point"})
```
Comment (kerthcet) [inline pkg/scheduler/metrics/metrics.go:198]: Maybe `profile` here is also needed?

**Comment by sanposhiho** [comment]
Thanks @kerthcet @alculquicondor.
These days, I'm running out of my time with KEPs related implementation, 
so will get back here after them. 🙏

File: pkg/scheduler/framework/runtime/framework.go (hunk 2 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:737]: Let's start with preFilter/filter.
@kidddddddddddddddddddddd will work on the prescore/score.
https://github.com/kubernetes/kubernetes/issues/110643#issuecomment-1430051794

I'm not sure if it's worth implementing on other extension points for now because it's other extension points don't have `Skip` feature. I'd say we can support other extension points when we support `Skip` on them.

File: pkg/scheduler/metrics/metrics.go (hunk 3 above)
Comment (sanposhiho) [reply pkg/scheduler/metrics/metrics.go:198]: Good point 👍

**Comment by sanposhiho** [comment]
@kerthcet @alculquicondor Sorry for being late on it. Fixed based on the suggestions:

- support PreFilter
- support profile label

**Comment by alculquicondor** [comment]
```
This metric counts how many times the specific plugin affects the scheduling result. In other words, the metric doesn't get incremented when the scheduling plugin makes no decision about an incoming Pod. For example, the metric only increments for the NodeAffinity plugin if a Pod being scheduled defines node affinity.
```

`Skip` is an implementation detail that users are not aware of.

File: pkg/scheduler/framework/runtime/framework.go [hunk 4]
```diff
@@ -734,6 +735,7 @@ func (f *frameworkImpl) RunFilterPlugins(
 		if state.SkipFilterPlugins.Has(pl.Name()) {
 			continue
 		}
+		metrics.PluginEvaluationTotal.WithLabelValues(pl.Name(), Filter, f.profileName).Inc()
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:738]: 🤔 why is the Filter value exported, but not preFilter? But this is outside of the scope of this PR.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

File: pkg/scheduler/metrics/metrics.go [hunk 5]
```diff
@@ -189,6 +189,14 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		}, []string{"plugin", "profile"})
 
+	PluginEvaluationTotal = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      SchedulerSubsystem,
+			Name:           "plugin_evaluation_total",
+			Help:           "Number of attempts to schedule pods by each plugin and the extension point (available only in PreFilter and Filter.).",
+			StabilityLevel: metrics.ALPHA,
+		}, []string{"plugin", "extension_point", "profile"})
```
Comment (logicalhan) [inline pkg/scheduler/metrics/metrics.go:198]: What's the cardinality of these labels?

Comment (alculquicondor) [reply pkg/scheduler/metrics/metrics.go:198]: <20, <10, <10

File: pkg/scheduler/framework/runtime/framework.go (hunk 4 above)
Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:738]: Seems `Filter` value is used in another package. (schedule_one.go)
https://github.com/sanposhiho/kubernetes/blob/799444a689da80470ad1aa87153f4feb5134b3db/pkg/scheduler/schedule_one.go#L533

It may be better to move those variables to metrics package? since they are used in the metric value's label.

**Comment by sanposhiho** [comment]
https://github.com/kubernetes/kubernetes/pull/115082#issuecomment-1456746354

modified the release not a bit

File: pkg/scheduler/framework/runtime/framework.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/runtime/framework.go:738]: yes, that sounds like a better location.

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:738]: PTAL https://github.com/kubernetes/kubernetes/pull/116312
